### DISCUSSION TITLE
Virtual Button Object / Collection Question

### DISCUSSION THREAD

**Discussion by r3sonant** [helgoboss/helgobox#456] [Q&A] [answered]
Hello.

I am trying to setup a controller preset for the X-Touch Mini.

I have mostly got it working but it's huge (amount of mappings, not file size) so I am wondering if I am doing it the best way.

What I am trying to configure is as follows.

8 rotary encoders.
Pressing the rotary encoder button switches to it's secondary instance.
Top row of illuminated buttons select banks 1 to 8 of rotary encoders.
1st button on bottom row switches between left and right sets of banks.

I am initially setting this up for Hive 2. I have seen the excellent presets by Vonglan but they didn't quite match how I wanted it to work, and I wanted to learn how to use the software properly for future synth mappings.

The way I am trying to set it up will have 192 mappings for the knobs alone though. Which seems excessive.

I was wondering if there was a way of creating a virtual encoder object? Where you would create a new instance, then add the encoder twice, and the button, then configure the button to switch between the two encoder states.

This could then be added to the main mapping as many times as needed and reduce the mapping count.

I want all the instance toggle buttons to be independent of each other, so if I have toggled to the second instance of button 1 in Bank 1, Left ,when I switch to bank 2 or Right I don't want them to be affected by this change. If I reused one toggle this didn't seem to be the case.

Is there a trick I am missing?

Any help would be appreciated.

Thanks.

**Comment by helgoboss** [comment]
> I was wondering if there was a way of creating a virtual encoder object? Where you would create a new instance, then add the encoder twice, and the button, then configure the button to switch between the two encoder states.
> 
> This could then be added to the main mapping as many times as needed and reduce the mapping count.

Can you describe this part in more detail or attach some screenshot, photo or RPP? What do you mean by "instance"?

You didn't speak about targets so far. Does all this happen in the controller compartment with targets that are either virtual or modify ReaLearn's internal parameters? If so, is your goal to create some kind of very complex virtual controller which you then want to reuse with different main presets?

**Comment by r3sonant** [comment]
Hello.

Thanks for the reply.

By instance I mean toggled state of the rotary encoder. So each encoder has two states (can be mapped to 2 different targets) which are switched between by pressing the encoder.

If it was possible to create a controller object in the controller mapping, then this object contained the encoder button and the encoder added twice. You would set it up in the object so the button toggled between the two states of the encoder.

Then when creating the main mappings you would just add the controller object to however many groups you needed, and they would all be separate instances, so changing values in one instance wouldn't effect another.

At the moment I have created 8 groups, and in each group created a mapping for each encoder and 2 mappings for each encoder button press, one for the left group in the Main mappings, and one for the right group.

![image](https://user-images.githubusercontent.com/11404006/136349171-b02a2f71-f93d-457e-83aa-ed76e4e8315b.png)

I have tried configuring the toggle parameters in the controller compartment but it didn't seem to work. It did when I did the same thing in the Main compartment.

Toggle Button Config:

![image](https://user-images.githubusercontent.com/11404006/136443038-72fb4a15-d849-487d-8b0f-fe57931c5617.png)

One of the toggled encoders config.

![image](https://user-images.githubusercontent.com/11404006/136443092-48e26ca2-ca9f-40e8-bc2c-147ebd5de014.png)

I have tried to create virtual mappings in the Controller Compartment for reuse in the Main Compartment but seem to have to keep adding new mappings in the controller compartment, then add them to the main compartment and configure them.

I did read through the manual and watch the video you made but have probably got completely the wrong end of the stick and have set it up all wrong.

But the goal is to create a controller mapping, which will be quite complex, that can then be used for different VSTis by creating a Main preset for each.

Thanks.

**Reply by helgoboss** [reply]
I don't know what you mean by *controller object* (this is not official ReaLearn vocabulary) but I suspect you want to do the following. Let's start simple, no bank buttons, no bank set buttons, easy to understand. Then you can figure out the rest by yourself.

- Your controller has 2 control elements which are push encoders:
    - e1
    - e2
- However, you want to make it look to the main compartment as if your controller actually had the following (virtual) control elements:
    - e1_not_pushed
    - e1_pushed
    - e2_not_pushed
    - e2_pushed

Is that correct, for starters?

**Reply by r3sonant** [reply]
That's exactly it.

The 'Controller Object' was just me thinking of OOP when trying to work out how best to configure ReaLearn for my own needs.

Thanks.

**Reply by helgoboss** [reply, marked as answer]
Import the following state from the clipboard.

- The solution is to use modifier-based conditional activation within the controller compartment (see user guide).
- I divided the controller mappings into 2 groups, one for each encoder. This is just for organizational purposes, it doesn't have any influence on functionality.
- For each encoder-turn message, you have 2 mappings: One that is only active if the push state is off (see *Active* field of mapping) and one that is only active if the push state is on. The push state for each encoder is stored as one of ReaLearn's internal parameters. Because "Parameter 1" and "Parameter 2" is not very descriptive, I gave them proper names: "Push state 1" and "Push state 2".
- For each encoder-push message, you have 1 mapping: It's a mapping that switches the corresponding "Push state" parameter on and off.

```json
{
  "version": "2.10.3",
  "id": "OISaP3_i",
  "controlDeviceId": "10",
  "feedbackDeviceId": "11",
  "defaultGroup": {},
  "defaultControllerGroup": {},
  "controllerGroups": [
    {
      "id": "89c072cc-95ab-4b1d-ad45-aeaf485a22f2",
      "name": "Push encoder 1"
    },
    {
      "id": "1942052d-34ff-468f-8f85-9d81fe4dda91",
      "name": "Push encoder 2"
    }
  ],
  "mappings": [
    {
      "id": "abb97cb9-cc2f-45e5-8463-6c3b5455b9e8",
      "name": "Example usage of non-pressed encoder 1",
      "source": {
        "category": "virtual",
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0,
        "controlElementIndex": "enc1/not-pressed"
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "type": 2,
        "trackGUID": "master",
        "fxAnchor": "id",
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      }
    },
    {
      "id": "d43224e5-d427-4643-affe-57381fa58610",
      "name": "Example usage of non-pressed encoder 2",
      "source": {
        "category": "virtual",
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0,
        "controlElementIndex": "enc1/pressed"
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "type": 4,
        "trackGUID": "master",
        "fxAnchor": "id",
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      }
    }
  ],
  "controllerMappings": [
    {
      "id": "a75fa46e-2430-4108-902d-48878bec9913",
      "name": "Push 1",
      "groupId": "89c072cc-95ab-4b1d-ad45-aeaf485a22f2",
      "source": {
        "channel": 1,
        "number": 0,
        "character": 1,
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "fxAnchor": "this",
        "paramIndex": 100,
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      }
    },
    {
      "id": "c18612a3-4c99-4791-9cef-f6a5b5270b60",
      "name": "Encoder 1 - not pressed",
      "groupId": "89c072cc-95ab-4b1d-ad45-aeaf485a22f2",
      "source": {
        "channel": 0,
        "number": 0,
        "character": 3,
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "category": "virtual",
        "fxAnchor": "id",
        "controlElementIndex": "enc1/not-pressed",
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      },
      "activationType": "modifiers",
      "modifierCondition1": {
        "paramIndex": 0,
        "isOn": false
      }
    },
    {
      "id": "757d1c47-a526-466b-a6c1-f53902aa6256",
      "name": "Encoder 1 - pressed",
      "groupId": "89c072cc-95ab-4b1d-ad45-aeaf485a22f2",
      "source": {
        "channel": 0,
        "number": 0,
        "character": 3,
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "category": "virtual",
        "fxAnchor": "id",
        "controlElementIndex": "enc1/pressed",
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      },
      "activationType": "modifiers",
      "modifierCondition1": {
        "paramIndex": 0,
        "isOn": true
      }
    },
    {
      "id": "5f8f582b-1647-4d17-b588-d8364d4717ee",
      "name": "Push 2",
      "groupId": "1942052d-34ff-468f-8f85-9d81fe4dda91",
      "source": {
        "channel": 1,
        "number": 1,
        "character": 1,
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "fxAnchor": "this",
        "paramIndex": 101,
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      }
    },
    {
      "id": "9d5b471b-1854-470e-bbd4-443a039e0870",
      "name": "Encoder 2 - not pressed",
      "groupId": "1942052d-34ff-468f-8f85-9d81fe4dda91",
      "source": {
        "channel": 0,
        "number": 1,
        "character": 3,
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "category": "virtual",
        "fxAnchor": "id",
        "controlElementIndex": "enc2/not-pressed",
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      },
      "activationType": "modifiers",
      "modifierCondition1": {
        "paramIndex": 1,
        "isOn": false
      }
    },
    {
      "id": "8c496e21-a5c9-406e-a1dd-4d4c50321de7",
      "name": "Encoder 2 - pressed",
      "groupId": "1942052d-34ff-468f-8f85-9d81fe4dda91",
      "source": {
        "channel": 0,
        "number": 1,
        "character": 3,
        "isRegistered": false,
        "is14Bit": false,
        "oscArgIndex": 0
      },
      "mode": {
        "maxStepSize": 0.05
      },
      "target": {
        "category": "virtual",
        "fxAnchor": "id",
        "controlElementIndex": "enc2/pressed",
        "useProject": true,
        "moveView": true,
        "seekPlay": true,
        "oscArgIndex": 0
      },
      "activationType": "modifiers",
      "modifierCondition1": {
        "paramIndex": 1,
        "isOn": true
      }
    }
  ],
  "controllerParameters": {
    "0": {
      "name": "Push state 1"
    },
    "1": {
      "name": "Push state 2"
    }
  }
}
```

**Comment by r3sonant** [comment]
Great, thank you.

I have imported it and will have a proper play later.

**Comment by r3sonant** [comment]
Out of curiosity, what's the best way to rename the parameters? I couldn't see any way of doing it through the GUI.

Do I just need to open the JSON file and change it there?

```
  "controllerParameters": {
    "0": {
      "name": "Push state 1"
    },
    "1": {
      "name": "Push state 2"
    }
  }
```

Thanks again.

**Reply by helgoboss** [reply]
Right mouse click in the header panel => Compartment parameters

**Comment by r3sonant** [comment]
Thank you.

Realearn is excellent (thanks again), but there is a learning curve...

**Reply by helgoboss** [reply]
There definitely is. Would be surprising if it wouldn't. After all, it's a sort of user-programmable general-purpose tool. That said, if you have some concrete suggestion (easy to implement if possible ;-) that would make user's life a bit easier, let me know.

**Comment by r3sonant** [comment]
I will do.

Thank you for all your help.
